src/lib/shim/src/tls.rs

```diff
@@ -128,179 +230,213 @@ impl From<FastThreadId> for usize {
+                // leak heap-allocated storage instead.
+                Box::leak(Box::new(StoragesType(core::array::from_fn(|_| {
+                    UnsafeCell::new(ShimThreadLocalStorage {
+                        _bytes: [0; BYTES_PER_THREAD],
+                    })
+                }))))
+            }
+        }
+    }
+
+    // Allocate lazily via `mmap`, to avoid unnecessarily consuming
     // the memory in processes where we always use native thread local storage.
     static STORAGES: LazyLock<&'static StoragesType> = LazyLock::const_new(|| {
-        let ptr = unsafe {
-            log::debug!("Allocating fallback thread-local storage");
-            rustix::mm::mmap_anonymous(
-                core::ptr::null_mut(),
-                core::mem::size_of::<StoragesType>(),
-                ProtFlags::READ | ProtFlags::WRITE,
-                MapFlags::PRIVATE,
-            )
-            .unwrap()
-        };
-        unsafe { &*(ptr as *const StoragesType) }
+        log::debug!("Allocating fallback thread-local storage");
+        StoragesType::alloc_new()
     });
 
+    /// Get the storage associated with `idx`. The bytes are initially zero.
     pub fn get(idx: usize) -> &'static UnsafeCell<ShimThreadLocalStorage> {
         &STORAGES.force().0[idx]
     }
 
+    /// Whether storage has been initialized.
     pub fn initd() -> bool {
         STORAGES.initd()
     }
 }
 
-// We can't do dynamic memory allocation via `mmap` under miri, so just
-// use statically allocated storage instead.
-#[cfg(miri)]
-mod global_storages {
-    use super::*;
-
-    static STORAGES: LazyLock<StoragesType> =
-        LazyLock::const_new(|| unsafe { core::mem::zeroed() });
-
-    pub fn get(idx: usize) -> &'static UnsafeCell<ShimThreadLocalStorage> {
-        let rv: &UnsafeCell<ShimThreadLocalStorage> = &STORAGES.force().0[idx];
-        // Unsafely extend the lifetime beyond the LazyLock reference.
-        unsafe { core::mem::transmute(rv) }
-    }
-
-    pub fn initd() -> bool {
-        STORAGES.initd()
-    }
-}
-
-// Should be called by every thread that accesses thread local storage, before
-// it makes the `exit` syscall and after it is done accessing thread local storage.
-//
-// This is a no-op when using native thread-local storage, but lets us reclaim memory
-// otherwise, and is necessary for correctness when `USE_FS_FALLBACK` is enabled.
-//
-// SAFETY: This thread must not access thread local storage again before exiting.
+/// Should be called by every thread that accesses thread local storage, before
+/// it makes the `exit` syscall and after it is done accessing thread local storage.
+///
+/// This is a no-op when using native thread-local storage, but lets us reclaim memory
+/// otherwise, and is necessary for correctness when [`Mode::NativeTlsId`] is enabled.
+///
+/// # Safety
+///
+/// This thread must not access thread local storage again before exiting.
```

Comment:
Maybe add ", and must not have any existing references to thread local storage" (I think just "must not access thread local storage" could make it easy to forget about things like drop impls or other methods of indirectly accessing thread locals)?